### PR TITLE
[ci] Add job for gappa

### PR DESCRIPTION
**Pull request by gares**
because it is in the windows installer

https://gitlab.inria.fr/gappa/coq/-/merge_requests/1 CC @silene 
conflicst docker of  #13464, to be merged after

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -795,6 +795,16 @@ library:ci-corn:
...
 
+plugin:ci-gappa:
+  extends: .ci-template-flambda
+  stage: stage-3
+  needs:
+  - build:edge+flambda
+  - library:ci-floq
+  dependencies:
+  - build:edge+flambda
+  - library:ci-floq
```
Comment (SkySkimmer) [inline .gitlab-ci.yml:806]:
```suggestion
  - library:ci-flocq
  dependencies:
  - build:edge+flambda
  - library:ci-flocq
```

**Comment by ejgallego** [comment]
There is an almost showstopper here, as gitlab.inria doesn't allow for us to submit overlays.

**Comment by gares** [comment]
I made a PR, but I don't think it is a usual overlay: the software without it does not compile even without this PR. This is why I did not make a proper overlay

**Review by ejgallego — DISMISSED** [review]
> There is an almost showstopper here, as gitlab.inria doesn't allow for us to submit overlays.

Turning this into a request for further discussion; I can live with ci developments that are in closed hosting services as long as they are Coq only [still a huge pain].

However, not being able to submit pull request for ML code seems like a much serious issue.

**Comment by gares** [comment]
@silene would you consider moving gappa to github/gitlab? The problem with *inria's* gitlab is that while *I* can fork/PR, to the best of my knowledge, you need an inria account to do that (I'm pretty sure it used to be the case, but maybe things changed).

**Comment by gares** [comment]
Aside from that, I don't get why this problem needs a discussion only now, and was not addressed when gappa was actually added to (windows) CI.

**Comment by ejgallego** [comment]
> Aside from that, I don't get why this problem needs a discussion only now, and was not addressed when gappa was actually added to (windows) CI.

Hard to recall, but I'm sure I was opposed to the current setup we have on Windows, but in the lack of manpower things moved on. I guess the argument was that as these developments were not tested in the CI it was not so problematic, and it was assumed the "??" would maintain that. Would be nice to recover the discussion as this is an example were we got things wrong.

Making life hard for the RM is already super bad indeed, but disrupting the regular overlay workflow is IMHO a very serious issue that we should carefully consider.

**Comment by silene** [comment]
> would you consider moving gappa to github/gitlab?

Not to Github. I am considering Gitlab, but not in the short term.

Also, to be clear, Inria's instance of Gitlab prevents from submitting issues and pull requests, but it does not prevent forks in any way. (It is Git, after all.) Thus, you can find a few forks named `gappa-coq` on Github, for example. Sure, it means you cannot click on a button on a webpage to perform a pull request. Instead, you have to use this awfully outdated means of communication called email.

Regarding this pull request itself, note that you don't need the Gappa tool and all its dependencies, if you are not running the testsuite. The Coq library and plugin depend on Flocq only, nothing else.

**Comment by gares** [comment]
Ok, thanks for the feedback.

What about a fork on younameit you watch and from time to time you push to? (A mirror of some sort). I wrote move, but we don't really care where the authoritative copy is as long the copy is not ignored by the upstream.

**Comment by gares** [comment]
I moved this PR on top of the other two https://github.com/coq/coq/pull/13464 https://github.com/coq/coq/pull/13481 that are currently open and change the docker, to avoid getting to do this later on.

I've also turned the overlay into a regular one.

Concerning the comment by @ejgallego this is my take after thinking about the problem:
- since anyone can fork gappa on github, make a PR there, and ping @silene or mail him, the extra complexity is on the upstream which has to merge on gitlab a PR prepared for github. I don't see the problem with this workflow.
- many of the core devs have an inria account, they can PR on the authoritative repo. These are the devs that create the majority of overlays for ML files, the kind that concerns @ejgallego  the most
- the upstream was made aware of the problem and is considering using a non-locked gitlab instance, solving the problem.
At the light of that, the problem seems more theoretical than practical to me.

**Comment by ejgallego** [comment]
The problem is not theoretical, as it could mean that if Gappa maintainers are on holidays our CI will stay broken for a long time.

This is a problem for other developments, but I would really like to have a fully automated workflow so overlays are submitted automatically by upcoming `submit-overlays.sh` and merged by coqbot, so we avoid the above problem.

IMHO we have two options:

- we use some mirror as official source for Gappa, then Gappa maintainers would have to pull from there when they want an update
- we move gappa to the Coq platform, and remove it from the installer in the CI. This seems like the proper place if it is not going to integrate with our CI

**Comment by gares** [comment]
> The problem is not theoretical, as it could mean that if Gappa maintainers are on holidays our CI will stay broken for a long time.

I don't see why this does not apply to Elpi for example, or MathComp, or Compcert... Maybe "theoretical" was badly chosen, but this discussion seems a bit off topic here.

> we use some mirror as official source for Gappa, then Gappa maintainers would have to pull from there when they want an update

This was an idea I also had in my pocket to the CUDW, but for all projects, not just Gappa. I think coqbot should be able to merge on the branch we track for CI on *all* project we track. This is a prerequisite, IMO, for the model we discussed yesterday, where coqbot (like bors) would do all merges (after ensuring a CI did fully run, *and* overlay merged). One way to achieve that is to host on github a mirror where coqbot has push rights on that branch (if the main repo is on github, then the branch can be the result of a PR from the authoritative repo).

**Comment by ejgallego** [comment]
> I don't see why this does not apply to Elpi for example, or MathComp, or Compcert... Maybe "theoretical" was badly chosen, but this discussion seems a bit off topic here.

I am assuming we can handle plugins and pure `.v` libraries differently, for plugins we don't offer backwards compatibility so that's a big deal, for `.v` libraries, we are not allowed to break compat except in special cases, so while the same issue exists, it is more minor IMHO so as I wrote above I can live with that for now [tho it is not optimal]

Elpi and other plugins are indeed hosted on open platfroms so at least PR submission can be automated now, integration with coqbot is to be discussed but I would think that those no willing to integrate, would be dropped from CI.


> This was an idea I also had in my pocket to the CUDW, but for all projects, not just Gappa. I think coqbot should be able to merge on the branch we track for CI on _all_ project we track.

This is indeed an option, IMO the choice should be down to the upstream plugin maintainer, but if it has to be that way, that's OK I think.

**Comment by SkySkimmer** [comment]
>This is a problem for other developments, but I would really like to have a fully automated workflow so overlays are submitted automatically by upcoming submit-overlays.sh and merged by coqbot, so we avoid the above problem.

We can just drop gappa when it happens and not worry about it prematurely.

**Comment by ejgallego** [comment]
> We can just drop gappa when it happens and not worry about it prematurely.

I'd rather not have that discussion happen when the first developer needs to submit an overlay to gappa and actually cannot open it.

It seems to me that we are doing here something similar to when we added Gappa to the Windows installer but not to the CI, thta worked fine, then later we got the problem. I'd like to keep the current overlay workflow working without all people submitting overalays having to be aware of this discussion and trying to fight with gitlab.inria and losing time. We have the platform , Gappa can be distributed from there.

**Comment by gares** [comment]
> then later we got the problem.

I had no problem making an overlay.

The problem I found is that the PR that did add gappa to windows did not add a job for gappa on linux (hence this PR) and the windows job does not build addons (by default) because it is slow.
As a consequence, 3 PR which would have needed an overlay for gappa did not have it, because CI did not detect it.

**Comment by Zimmi48** [comment]
I'm on the side of @gares and @SkySkimmer for this PR. OTOH, @gares, you don't absolutely need this in for 8.13, do you? If you actually don't, we could delay the merge until this is discussed during CUDW or a Coq Call to not force @ejgallego's hand.

**Comment by MSoegtropIMC** [comment]
Shouldn't we move CI to use opam, possibly with a local patch repo as I have it for the platform? I found that Opam is quite flexible.

**Comment by ejgallego** [comment]
> I'm on the side of @gares and @SkySkimmer for this PR. OTOH, @gares, you don't absolutely need this in for 8.13, do you? If you actually don't, we could delay the merge until this is discussed during CUDW or a Coq Call to not force @ejgallego's hand.

If you folks are OK with this I'm fine; I remain worried about the overlay process tho, what do you propose when we have to add an overlay for Gappa?

**Comment by ejgallego** [comment]
To be more concrete, would we merge this, I'm unsure what to do with overlays involving gappa anymore:
- if I am the submitter, I cannot open a PR to gappa upstream, so what should I do? Remove gappa? Fork gappa?
- if I am the assignee, and I face problems with inria.gitlab, what should I do?
- moreover, this block the planned automation in both submission and merge time

I worry a bit more that if I am having this problem myself, what kind of problems other devs will face? We have recently a core dev lose a lot of time due to CI, so I'd rather not introduce situations that could lead to this kind of loss of time.

**Comment by gares** [comment]
>     * I cannot open a PR to gappa upstream

You know that *you* can right?

**Comment by ejgallego** [comment]
> You know that _you_ can right?

I can, but not all devs can.

**Comment by gares** [comment]
This is what I would do to propose an overlay for gappa if I had no Inria account:
```shell
. dev/ci/ci-basic-overlay.sh
git clone -o upstream $gappa_plugin_CI_GITURL -b $gappa_plugin_CI_REF gappa-coq
cd gappa-coq
git remote add origin git@github.com:gares/gappa-coq.git
git checkout -b my-fix
# hack
git push origin my-fix
# follow the link printed by github, open the PR, write @silene in the message.
```
The only differences with what I had to do to submit the overlay in this PR are:
- the remote for origin had a different URL
- I did not have to write `@silene` since it got automatically notified

As an assignee, at merge time, I would go to the same PR and write `@silene please merge`, no difference.

**Comment by SkySkimmer** [comment]
>follow the link printed by github

Is there a link?

**Comment by gares** [comment]
It prints the link to open the pr with the branch you pushed

**Comment by ejgallego** [comment]
The discussion here made me think that a few problems we are having is that indeed we don't yet have a fully coherent strategy in regards of how we see tooling. For example, when we started the CI, my key goal [that we have put a lot of effort into] was to have:

- a fully integrated dev setup, with a composed incremental build, that would allow to exactly reproduce the CI / benchs locally
- fully "automated" management of the multi-repository setup [branching, merging, etc...]
- a common standard package setup/build _à la cargo_ if you wish 
- reusing a lot of the OCaml goals / strategy / tooling as to easy maintenance [in this case for example opam cannot met some of the goals stated above]
- avoiding diverging upstream given the experience with coq-contribs

The main motivation for the above set of requirements was to allow a quick development pace, and in particular foster a cleanup/modernization of Coq's codebase as this is a key point in order to have a more modular and maintanable base on which build the new research features many people want / work on. I do think that's an important motivation, see for example Lean, which was in large part motivated as to have a fresh implementation of what basically is Coq's TT.

I would be happy to abandon this goal if we think it is not the good one; there are alternatives, and that's OK. But indeed, IMHO we should advance towards a better specification of what we want from our tooling setup, and try to be all in sync so we don't do parallel / incompatible work.

The above strategy is indeed making some choices in lot of stuff, for examples, packages hosted in exotic services or using non-standard build systems cannot get included in the CI; this is IMHO the right choice as we can't expect devs to have to know every possible variation implemented by such packages, but indeed, there is a tradeoff, where I naturally sit on the side of the Coq developer wanting to have a standard and fully integrated build so I don't have to do manual steps in my day-to-day dev.

**Comment by gares** [comment]
I think this is enough. This discussion has no place here. We do have CoqCalls, CEPs, CUWD workshops...

Coq 8.12.x, that you and Theo released, includes these addons: (gappa is there, I save you the time looking for it)
https://github.com/coq/coq/blob/79c0e0d2e238d604c043749aebcd6ccf2388953f/dev/ci/gitlab.bat#L42-L60
The only thing I want to do, is to be able to release the same installer with the help of CI. This is what this PR is for.

Please review this PR on the basis of that, and discuss all the rest in the most appropriate venue.
Everything can be changed, improved and even reverted.

If you continue to stall this PR, I'll merge it directly in v8.13 myself. I'm fine with that, no hard feelings. I don't need to merge this in master. Having it in master may help the next release manager to not discover breakage late, as I did. But It's not needed, strictly speaking, for 8.13, which I'm responsible for, and which lives on his branch already.

**Comment by ejgallego** [comment]
I am not stalling this PR, read again what I wrote please.

And indeed if this is not the place to discuss this [relative to master, of course, 8.13 indeed should get the fix now] , I'm not sure which one it is. The moment we merge this the dev workflow will be disrupted, but I already explained that so I will save everyone of repeating myself.

Also note that the part relevant for the installer can be also merged right now in `master`, there could well be 2 separate PRs.

**Comment by Zimmi48** [comment]
My take on this is that the goals you stated @ejgallego are still valid, but they are not seriously hindered by this PR, which only *fixes* a situation that should have never happened (Windows add-ons without corresponding CI).

We have already a problem with the multiplicity of platforms used by projects that we test (and Inria's GitLab is already one of them, another one being MPI's GitLab on which each overlay submitter needs to manually request access). Clearly, this means that we should focus on providing better tooling for submitting overlays instead of debating on a PR that is fundamentally *useful*. There is a simple solution to make overlay submission straightforward on all platforms: it is to delegate to @coqbot the task to open the PRs. Let's focus on achieving a super simplified workflow instead of losing time on this PR. And let's move the discussion elsewhere (e.g., in https://github.com/coq/bot/issues/72 or a new issue) instead of highjacking this PR.

**Comment by gares** [comment]
I've removed the last commit, since (a variant of) the overlay was merged in master.

**Comment by Zimmi48** [comment]
Spurious async failure that occurs from time to time.

**Comment by ejgallego** [comment]
> My take on this is that the goals you stated @ejgallego are still valid, but they are not seriously hindered by this PR, which only _fixes_ a situation that should have never happened (Windows add-ons without corresponding CI).
> 
> And let's move the discussion elsewhere (e.g., in [coq/bot#72](https://github.com/coq/bot/issues/72) or a new issue) instead of highjacking this PR.

Note that my point is that there are two possible solutions to the problem here, one is adding to the CI, the other is removing the addon from the installer. But fair enough, we can move the discussion to coqbot, tho this still impacts coq IMO. As I wrote above "I'm fine merging this if you folks think it's fine" [I used the review button as otherwise discussion tends not to happen]

> We have already a problem with the multiplicity of platforms used by projects that we test (and Inria's GitLab is already one of them, another one being MPI's GitLab on which each overlay submitter needs to manually request access). 

MPI Gitlab is used for pure .v developments, IMO that's less of an issue. My fear with this is reading ML-level overlays. So indeed we can go ahead, but the concrete fear I expressed above still persists [some developer wondering why we went that way when submission of overlays is disrupted]. So indeed I'm not sure I can add anything else other than if the situation happens, we'll have to tell the devs that indeed they have to rely on a proxy to submit gappa's overlays.

**Comment by SkySkimmer** [comment]
Needs rebase for the cachekey stuff

**Review by SkySkimmer — APPROVED** [review]
>So indeed we can go ahead

OK let's go

**Comment by SkySkimmer** [comment]
@coqbot: merge now

**Pull request merged**